frame/grandpa/src/lib.rs

```diff
@@ -189,10 +211,81 @@ decl_module! {
 
 		fn deposit_event() = default;
 
-		/// Report some misbehavior.
-		fn report_misbehavior(origin, _report: Vec<u8>) {
-			ensure_signed(origin)?;
-			// FIXME: https://github.com/paritytech/substrate/issues/1112
+		/// Report voter equivocation/misbehavior. This method will verify the
+		/// equivocation proof and validate the given key ownership proof
+		/// against the extracted offender. If both are valid, the offence
+		/// will be reported.
+		///
+		/// FIXME: I have no clue about the weight (but we're checking two
+		/// ed25519 signatures).
+		#[weight = SimpleDispatchInfo::FixedOperational(10_000_000)]
+		fn report_equivocation(
+			origin,
+			equivocation_proof: EquivocationProof<T::Hash, T::BlockNumber>,
+			// key_owner_proof: <T::HandleEquivocation as equivocation::HandleEquivocation<T>>::KeyOwnerProof,
```

Comment:
I don't know how to fix this. @bkchr mentioned that this is probably some rustc bug.

Reply:
This is what I did in the initial implementation, the problem is that it requires always having a `KeyOwnerProofSystem` in order to instantiate GRANDPA (e.g. node-template doesn't use any of this), so I refactored all that into the `HandleEquivocation` trait that we can implement for `()` in order to equivocation reporting. We can't implement `KeyOwnerProofSystem` for `()` we'd need something like struct `NoopKeyOwnerProofSystem<Key, Proof, Id>(PhantomData<(Key, Proof, Id)>)`.

Reply:
No longer an issue since I moved `KeyOwnerProof` to the module trait and there's no extra indirection now.